src/caffe/util/gpu_memory.cpp

```diff
@@ -0,0 +1,186 @@
+#include <algorithm>
+#include <vector>
+#include "caffe/common.hpp"
+#include "caffe/util/gpu_memory.hpp"
+
+#ifndef CPU_ONLY
+#include "cub/util_allocator.cuh"
+#endif
+
+namespace caffe {
+
+#ifndef CPU_ONLY
+static cub::CachingDeviceAllocator* cub_allocator = 0;
+vector<GPUMemoryManager::MemInfo> GPUMemoryManager::dev_info_;
+#endif
+
+GPUMemoryManager::PoolMode GPUMemoryManager::mode_ = GPUMemoryManager::NO_POOL;
+bool GPUMemoryManager::debug_ = false;
+
+#ifdef CPU_ONLY
+void GPUMemoryManager::init(const std::vector<int>&, PoolMode, bool) {}
+void GPUMemoryManager::destroy() {}
+
+const char* GPUMemoryManager::pool_name() {
+  return "No GPU: CPU Only Memory";
+}
+#else
+
+void GPUMemoryManager::init(const std::vector<int>& gpus, PoolMode m,
+    bool debug) {
+  bool debug_env = getenv("DEBUG_GPU_MEM") != 0;
+  debug_ = debug || debug_env;
+  if (gpus.size() <= 0) {
+    m = GPUMemoryManager::NO_POOL;
+  }
+
+  switch (m) {
+  case CUB_POOL:
+    InitMemory(gpus, m);
+    break;
+  default:
+    break;
+  }
+  VLOG_IF(1, debug) << "GPUMemoryManager initialized with " << pool_name();
+}
+
+void GPUMemoryManager::destroy() {
+  switch (mode_) {
+  case CUB_POOL:
+    delete cub_allocator;
+    cub_allocator = NULL;
+    break;
+  default:
+    break;
+  }
+  mode_ = NO_POOL;
+}
+
+void GPUMemoryManager::allocate(void** ptr, size_t size, cudaStream_t stream) {
+  CHECK((ptr) != NULL);
+  switch (mode_) {
+  case CUB_POOL:
+    if (cub_allocator->DeviceAllocate(ptr, size, stream) != cudaSuccess) {
+      int cur_device;
+      CUDA_CHECK(cudaGetDevice(&cur_device));
+      // free all cached memory (for all devices), synchrionize
+      cudaDeviceSynchronize();
+      cudaThreadSynchronize();
+      cub_allocator->FreeAllCached();
+      cudaDeviceSynchronize();
+      cudaThreadSynchronize();
+      // Refresh per-device saved values.
+      for (int i = 0; i < dev_info_.size(); ++i) {
+        // only query devices that were initialized
+        if (dev_info_[i].total_) {
+          update_dev_info(i);
+          // record which device caused cache flush
+          if (i == cur_device) {
+            dev_info_[i].flush_count_++;
+          }
+        }
+      }
+      // Retry once
+      CUDA_CHECK(cub_allocator->DeviceAllocate(ptr, size, stream));
+    }
+    // If retry succeeds we need to clean up last error
+    cudaGetLastError();
+    break;
+  default:
+    CUDA_CHECK(cudaMalloc(ptr, size));
+    break;
+  }
+}
+
+void GPUMemoryManager::deallocate(void* ptr, cudaStream_t stream) {
+  // allow for null pointer deallocation
+  if (!ptr) {
+    return;
```

Comment:
Being able to deallocate a null pointer is consistent with `cudaFree`, `free`, and `delete`.